feat: configurable default views

hrwX opened the pull request:
- The views for a doctype are not configurable by the user, ie if I want the framework to always route the user to the kanban view by default, the framework doesn't allow that.
- This PR tries to address that by adding a configurable selector for the view to be displayed.
- The doctype won't default to ListView but route to the default view from meta.
- This will remove the need for maintaining `treeviews` and `calendars` in the boot, and will be handled directly from doctype meta.
- ![image](https://user-images.githubusercontent.com/7310479/174721419-bc10c443-edfd-48cf-b854-7fcf2f104ea4.png)
- ![image](https://user-images.githubusercontent.com/7310479/174721793-de9c9292-9e80-4c36-856e-a16a1a39cbd4.png)

https://docs.erpnext.com/docs/v13/user/manual/en/customize-erpnext/customize-form/edit?wiki_page_patch=194596b062

hrwX committed:
feat: configurable default views

hrwX committed:
feat: configurable default views

hrwX committed:
chore: sider fixes

hrwX committed:
fix: show force reroute check

barredterra commented on frappe/custom/doctype/customize_form/customize_form.js:
```diff
@@ -233,6 +238,39 @@ frappe.ui.form.on("Customize Form", {
+	},
+
+	setup_default_views(frm) {
+		frappe.model.with_doctype(frm.doc.doc_type, () => {
```
Code duplication from `doctype.js`. Can we factor this out into a utility?

barredterra reviewed (changes requested):
When "Force reroute" is enabled
- [x] disable the view switcher in List View, etc.

    User doesn't need to see a button "Report View" if he will be redirected to Kanban view anyways.

- [x] change the URL to the correct (forced) route

    For example, if I'm rerouted from report to list, the url still shows "report" at the moment.

For Kanban view, we might need to specify a board. Otherwise, it doesn't seem to work.

hrwX commented:
> * e a button "Report View" if he will be redirected to Kanban view anyw

- disable the view switcher in List View, etc.
    -  Even if the user is forcefully redirected to ReportView, the user should be able to switch to a different view, removing the switcher could kinda handicap the user. For eg: for calendar doctype, if the default view is set to report, the user wont be able to switch to the calendar view.

- change the URL to the correct (forced) route: 
    - fixing it

- or Kanban view, we might need to specify a board. Otherwise, it doesn't seem to work.
    - will figure out a way to make this work

barredterra commented:
> Even if the user is forcefully redirected to ReportView, the user should be able to switch to a different view

He won't be able to switch, because he'll be forcefully redirected. So the button becomes useless. 😄

hrwX committed:
fix: handle kanban board rendering

hrwX committed:
fix: show columns for kanban

hrwX commented:
> > Even if the user is forcefully redirected to ReportView, the user should be able to switch to a different view
> 
> He won't be able to switch, because he'll be forcefully redirected. So the button becomes useless. smile

- makes sense, pushed the change

hrwX committed:
chore: fix folder navigation tests

hrwX committed:
fix: do not reroute for File doctype

hrwX committed:
chore: run pre-commit

hrwX committed:
test: add test cases

hrwX committed:
chore: change doctype to event

ankush commented:
related UI test failing

<img width="861" alt="Screenshot 2022-07-07 at 11 20 57 AM" src="https://user-images.githubusercontent.com/9079960/177701307-995c22d4-bbd2-4b81-854e-b127644afcc8.png">

hrwX committed:
test: change doctype to event

hrwX commented:
- Unrelated tests failing 
- ![image](https://user-images.githubusercontent.com/7310479/177992962-d4f101a4-4ef1-4dcd-a0a0-65b767f409fa.png)

barredterra commented:
@hrwX, with _force re-route_ enabled, everything works as expected. 🎉 

Without _force re-route_, the behavior is a bit weird now. For example:

1. open DocType **Task**
2. set default view to "Report", and _force re-route_ to `0`
3. reload the page
4. go to `/app/task` -> seeing the list view 😕
5. click around the system a bit
6. go to the `/app/projects` workspace and click on "Task" shortcut -> seeing tree view 😕 

Also, I'm not sure if the first kanban (according to which order?) is always the best kanban. Maybe also show the dialog, if there's more than one?

hrwX commented:
> @hrwX, with _force re-route_ enabled, everything works as expected. tada
> 
> Without _force re-route_, the behavior is a bit weird now. For example:
> 
> 1. open DocType **Task**
> 2. set default view to "Report", and _force re-route_ to `0`
> 3. reload the page
> 4. go to `/app/task` -> seeing the list view confused
> 5. click around the system a bit
> 6. go to the `/app/projects` workspace and click on "Task" shortcut -> seeing tree view confused
> 
> Also, I'm not sure if the first kanban (according to which order?) is always the best kanban. Maybe also show the dialog, if there's more than one?


@barredterra 
- The default view will only come into play if the no specific route is provided when routing, that is when the default view would be shown. If that behaviour would be changed, the user would not be able to access the Report/List/Kanban.
- The task one is really weird, I am debugging that as of now

barredterra commented:
@hrwX I agree, the specific `/app/task/view/list` should not get redirected to `/app/task/view/report`, if _force reroute_ is disabled. But the unspecific `/app/task` should get redirected to the configured default view `/app/task/view/report`. The latter didn't seem to work for me yet.

hrwX committed:
fix: routing to default view and kanban selector

hrwX commented:
> @hrwX I agree, the specific `/app/task/view/list` should not get redirected to `/app/task/view/report`, if _force reroute_ is disabled. But the unspecific `/app/task` should get redirected to the configured default view `/app/task/view/report`. The latter didn't seem to work for me yet.

@barredterra I have fixed the routing for `app/{doctype}/view` and added a kanban selector.
![Peek 2022-07-15 12-06](https://user-images.githubusercontent.com/7310479/179212093-17ada125-cfcb-4325-a6aa-1f630dccddf4.gif)

![Peek 2022-07-15 12-11](https://user-images.githubusercontent.com/7310479/179212488-207985f6-ff6a-4ba6-af12-b953dbfb001d.gif)

hrwX committed:
Update view_routing.js

barredterra reviewed (commented):
The default view, without _Force Re-route to Default View_ enabled, still doesn't have any effect for me.

I tried the following:

- Open **Customize Form** for **Item**, set _Default View_ to "Image". Visiting `/app/item` still shows the list view.
- Open **Customize Form** for **Task**, set _Default View_ to "Kanban". Visiting `/app/task` still shows the list view.

hrwX committed:
fix: reroute to list for app/{doctype} route

hrwX commented:
> The default view, without _Force Re-route to Default View_ enabled, still doesn't have any effect for me.
> 
> I tried the following:
> 
> * Open **Customize Form** for **Item**, set _Default View_ to "Image". Visiting `/app/item` still shows the list view.
> * Open **Customize Form** for **Task**, set _Default View_ to "Kanban". Visiting `/app/task` still shows the list view.

@barredterra 
Now the route `app/{doctype}` and `app/{doctype}/view` will route based on default view

barredterra reviewed (commented):
@hrwX functionality works as expected now 🥳 . Please check the CI failures.

hrwX committed:
chore: sider fixes

hrwX committed:
chore: linter changes

hrwX committed:
chore: linter changes

hrwX committed:
fix: do not reroute unnecessarily

hrwX committed:
chore: route to view/home for file tests

hrwX committed:
fix: back button popstate

hrwX committed:
fix: handle file view edgecase and test cases

barredterra commented on frappe/tests/ui_test_helpers.py:
```diff
@@ -333,3 +349,86 @@ def insert_child(doc, data, barcode, check, rating, duration, date):
+
+
+@frappe.whitelist()
+def setup_tree_doctype():
+	frappe.delete_doc_if_exists("DocType", "Custom Tree")
```
Do we need to check for permission or `frappe.flags.in_test` in these whitelisted methods??

hrwX replied:
I don't think so it would be necessary since I am not changing any existing doctype, just adding a few for testing in cypress

hrwX committed:
chore: linter changes

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
fix: file view nested routing

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
chore: do not add last_name for admin in tests

hrwX committed:
test: fix routes in tests

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
chore: kanban test fixes

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
chore: insert kanban only if not exists

hrwX committed:
chore: check kanban for note

hrwX committed:
chore: fileview fixes

hrwX committed:
chore: change selector for ui test

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

barredterra commented:
Configure Kanban as default view for the **Task** DocType:
![Bildschirmfoto 2022-08-11 um 17 12 22](https://user-images.githubusercontent.com/14891507/184167660-4c30df2d-20fe-4345-8b87-8ccef3a3fae2.png)

Open `/app/task`, click on "Go To Task List" in DocType, or open **Task** from the "Projects" Workspace:

![Bildschirmfoto 2022-08-11 um 17 12 06](https://user-images.githubusercontent.com/14891507/184167973-a44884ba-995a-4a67-9666-9a3112d1abef.png)

barredterra commented:
1. Configuring default view "Image" for **Item** and enabling "Force reroute"
2. Open `/app/item/view/report`
3. Shows image view correctly, but the URL stays wrong (`/app/item/view/report`)

hrwX committed:
chore: remove load_last_view for kanban

hrwX commented:
> Configure Kanban as default view for the **Task** DocType: ![Bildschirmfoto 2022-08-11 um 17 12 22](https://user-images.githubusercontent.com/14891507/184167660-4c30df2d-20fe-4345-8b87-8ccef3a3fae2.png)
> 
> Open `/app/task`, click on "Go To Task List" in DocType, or open **Task** from the "Projects" Workspace:
> 
> ![Bildschirmfoto 2022-08-11 um 17 12 06](https://user-images.githubusercontent.com/14891507/184167973-a44884ba-995a-4a67-9666-9a3112d1abef.png)

This is fixed now, `load_last_view` for kanban was the problem

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
fix: re reoute to default view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
chore: reintroduce load_last_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
fix: file view routing

hrwX committed:
fix: set Home as default filter for fileview

barredterra commented:
@hrwX
 
https://user-images.githubusercontent.com/14891507/184493117-18b906a7-9110-408f-90b9-d36ed2eb4303.mov

barredterra committed:
Merge branch 'develop' into default_view

hrwX committed:
fix: kanban view for shortcut widget

barredterra reviewed (commented):
Tested, works as expected. However, related Cypress tests seem to be failing.

hrwX committed:
Merge branch 'develop' into default_view

dufani1 commented:
Very nice feature

barredterra committed:
Merge branch 'develop' into default_view

barredterra committed:
test: re-whitelist helper method

Was accidentally removed

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

hrwX committed:
Merge branch 'develop' into default_view

barredterra commented:
Failing test seems to be unrelated

hrwX committed:
Merge branch 'develop' into default_view

dufani1 commented:
Please merge this feature... it is very helpful

hrwX committed:
Merge branch 'develop' into default_view

barredterra committed:
Merge branch 'develop' into default_view

barredterra reviewed (dismissed):
Failing tests seem to be unrelated.

ankush committed:
Merge branch 'develop' into default_view

barredterra committed:
Merge branch 'develop' into default_view

shariquerik commented:
Merged it by mistake and also reverted it has raised a separate PR https://github.com/frappe/frappe/pull/18409
